cyraft/node.py

```diff
@@ -451,6 +451,51 @@ async def _serve_execute_command(
             return uavcan.node.ExecuteCommand_1.Response(uavcan.node.ExecuteCommand_1.Response.STATUS_SUCCESS)
         return uavcan.node.ExecuteCommand_1.Response(uavcan.node.ExecuteCommand_1.Response.STATUS_BAD_COMMAND)
 
+    def _reset_election_timeout(self) -> None:
+        _logger.info("Node ID: %d -- Resetting election timeout", self._node.id)
+        loop = asyncio.get_event_loop()
+        self._next_election_timeout = loop.time() + self._election_timeout
+        self._election_timer.cancel()
+        self._election_timer = loop.call_at(self._next_election_timeout, self._on_election_timeout)
+
+    def _on_election_timeout(self) -> None:
+        if self.state == RaftState.FOLLOWER or self.state == RaftState.CANDIDATE:
+            _logger.info("Node ID: %d -- Election timeout reached", self._node.id)
+            self.prev_state = self.state
+            self.state = RaftState.CANDIDATE
+            loop = asyncio.get_event_loop()
+            loop.call_soon_threadsafe(asyncio.create_task, self._start_election())
+            # loop.call_soon_threadsafe(self._start_election)
+            self._reset_election_timeout()
```

Comment:
Is the usage of `loop.call_soon_threadsafe` correct here?

Initially I wanted to use it in the commented out way, however this would result in the following error:

![Screenshot 2023-05-07 at 16 17 54](https://user-images.githubusercontent.com/39975120/236680245-c4cb3125-e5fa-4784-9882-34f0561c6162.png)

(Adding additional sleep time wouldn't solve this)

The reason I'm asking: in `pycyphal` it's done in the commented out manner, so I'm suspecting I'm missing something here.

(The above fix was suggested by ChatGPT, however PavelGPT seems to be more reliable in these delicate matters)

Reply:
>in pycyphal it's done in the commented out manner, so I'm suspecting I'm missing something here.

Can you please point me to that code? You are correct that it is meaningless to call a coroutine without awaiting it.

Most importantly, I don't see why do you need to use the thread-safe postponed execution here. If I understand your code correctly, all of it runs in the same thread (but different coroutines, which is fine), so you should be able to just do this:

```python
await self._start_election()
```

Reply:
Starting a new coroutine via `asyncio.create_task(self._start_election)` is also fine but it creates complications with error handling. You can't just ignore the fact that exceptions may be thrown from the coroutine, they have to be handled & propagated to the caller correctly, and this is where it may get a bit hairy.

Reply:
> Can you please point me to that code? You are correct that it is meaningless to call a coroutine without awaiting it.

For example [here](https://github.com/OpenCyphal/pycyphal/blob/4f56e6518194268c7a6ae6c4b232e5a7aead7005/pycyphal/transport/udp/_session/_input.py#L179)

![image](https://user-images.githubusercontent.com/39975120/236767791-68e49ba7-fbf7-4e33-83a4-fd5396472650.png)

Reply:
> I don't see why do you need to use the thread-safe postponed execution here.

If I try to use `await`, it means that `_on_election_timeout` becomes an async function.

`loop.call_at(...)` expects a regular function, otherwise I get the following error:

![image](https://user-images.githubusercontent.com/39975120/236769840-4b88dffc-c38f-43fd-bf50-885df401b26f.png)

Tried to fix by using an additional wrapper function, however still same issue remains. 

See commit "trying to use await"